contracts/pcv/EthPCVSwapperUniswap.sol

```diff
@@ -0,0 +1,145 @@
+    /// @notice the maximum amount of tokens to buy
+    uint256 public tokenBuyLimit;
+    /// @notice the maximum amount of ether to spend on every swap
+    uint256 public maxEthSpentPerSwap;
+    /// @notice the maximum amount of slippage vs oracle price
+    uint256 public maximumSlippage = 300; // default 3%
+    uint256 public constant SLIPPAGE_GRANULARITY = 10000;
+
+    constructor(
+        address _core,
+        address _pair,
+        address _router,
+        address _oracle,
+        uint256 _swapFrequency,
+        address _tokenReceived,
+        uint256 _tokenBuyLimit,
+        uint256 _maxEthSpentPerSwap
+    ) public UniRef(_core, _pair, _router, _oracle) Timed(_swapFrequency) {
+        tokenReceived = _tokenReceived;
+        tokenBuyLimit = _tokenBuyLimit;
+        maxEthSpentPerSwap = _maxEthSpentPerSwap;
+
+        emit UpdateTokenSpent(router.WETH());
+        emit UpdateTokenReceived(_tokenReceived);
+
+        // start timer
+        _initTimed();
+    }
+
+    receive() external payable {}
+
+    /// @notice withdraw ETH from the reserves
+    /// @param to address to send ETH
+    /// @param amountOut amount of ETH to send
+    function withdrawETH(address payable to, uint256 amountOut) external override onlyPCVController {
+        Address.sendValue(to, amountOut);
+        emit WithdrawETH(msg.sender, to, amountOut);
+    }
+
+    /// @notice withdraw ERC20 from the reserves
+    /// @param to address destination of the ERC2°
+    /// @param token address of the ERC20 to send
+    /// @param amount amount of ERC20 to send
+    function withdrawERC20(address to, address token, uint256 amount) external override onlyPCVController {
+        require(IERC20(token).transfer(to, amount), "EthPCVSwapperUniswap: transferERC20 failed");
+        emit WithdrawERC20(msg.sender, to, token, amount);
+    }
+
+    /// @notice Sets the token to spend
+    /// @param _tokenSpent the address of the token to spend
+    function setTokenSpent(address _tokenSpent) external override onlyGovernor {
+        require(_tokenSpent == router.WETH(), "EthPCVSwapperUniswap: can only spend ETH");
+    }
+
+    /// @notice Sets the token to receive
+    /// @param _tokenReceived the address of the token to receive
+    function setTokenReceived(address _tokenReceived) external override onlyGovernor {
+      tokenReceived = _tokenReceived;
+      emit UpdateTokenReceived(_tokenReceived);
+    }
+
+    /// @notice Get the token to spend
+    /// @return The address of the token to spend
+    function getTokenSpent() external view override returns (address) {
+      return router.WETH();
+    }
+
+    /// @notice Get the token to receive
+    /// @return The address of the token to receive
+    function getTokenReceived() external view override returns (address) {
+      return tokenReceived;
+    }
+
+    function swap() external override afterTime whenNotPaused {
+      require(address(this).balance > 0, "EthPCVSwapperUniswap: no ETH left");
```

Comment:
If we switch to using ERC20s we'd need to change this logic. You could also require != instead of > which saves a tiny bit of gas

Reply:
Updated with a `IERC20(tokenSpent).balanceOf(address(this))` instead of `address(this).balance` & changed to `!=`